### PR TITLE
Add `torch.count_nonzero`

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference #38349 

TODO:

* [x] Add tests
* [x] Add docs (pending add to docs.rst)

### CODE REVIEW

**Comment by kshitij12345** [comment]
@mruberry 

Please review

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -805,6 +805,14 @@
 - func: cosine_embedding_loss(Tensor input1, Tensor input2, Tensor target, float margin=0.0, int reduction=Mean) -> Tensor
   use_c10_dispatcher: full
 
+- func: count_nonzero.dim_IntList(Tensor self, int[] dims) -> Tensor
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:808]: Might as well make these methods, too.

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -1636,6 +1636,31 @@ def merge_dicts(*dicts):
             -1.8209, -2.9780, -3.4022])
 """.format(**reduceops_common_args))
 
+add_docstr(torch.count_nonzero,
+           r"""
+count_nonzero(input, dim=None) -> Tensor
+
+Counts the number of non-zero values in the tensor :attr:`input`.
```
Comment (mruberry) [inline torch/_torch_docs.py:1643]: This should explain the dim argument, too.

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -1636,6 +1636,31 @@ def merge_dicts(*dicts):
...
+
+Args:
+    {input}
+    dim  (int or tuple of ints, optional): Dim or tuple of dims along which to count non-zeros.
+    Default is None, meaning that non-zeros will be counted along a flattened version of :attr:`input`.
+
+Example::
+
+    >>> x = torch .zeros(3,3)
```
Comment (mruberry) [inline torch/_torch_docs.py:1652]: You have a space between torch and .zeros

File: test/test_torch.py [hunk 4]
```diff
@@ -17596,6 +17596,28 @@ def test_multinomial_empty(self, device):
...
+    @dtypes(*(torch.testing.get_all_int_dtypes() + torch.testing.get_all_fp_dtypes(include_bfloat16=False)))
+    def test_count_nonzero(self, device, dtype):
+        # Test for default `dim=None`
+        rand_dim = random.randint(3, 4)
+        shape = self._rand_shape(rand_dim, 5, 10)
+        x = torch.randn(*shape)
+        x[torch.randn(*shape) > 0.5] = 0
+        self.compare_with_numpy(torch.count_nonzero, np.count_nonzero, x, device, dtype)
+
+        rand_dim = random.randint(3, 4)
```
Comment (mruberry) [inline test/test_torch.py:17609]: This test is pretty good but I think it can be simplified. What about a loop like:

```
for ndims in (0, 1, 3):
  # Create tensor with ndims
  for n in range(ndims + 1):
      for c in itertools.combinations(list(range(ndims)), n):
        for p in itertools.permutations(c):
          # Do test case
```

That should test every dim combination for 0D, 1D, and 3D tensors and be easy to extend, if needed, in the future. You'll have to test for `()` and use that as a signal not to pass a dim argument.

**Review by mruberry — CHANGES_REQUESTED** [review]
Hey @kshitij12345, thanks for the PR! I added some comments. 

Summary + followup checklist: 

- Suggestion to simplify test in test_torch.py
- Minor doc improvements
- Method variants
- Add a "not_implemented" entry to tools/autograd/derivatives.yaml
- Update docs/source/name_inference.rst
- Update docs/source/tensors.rst and docs/source/torch.rst
- Update torch/_torch_docs.py

Overall this is really good, it just needs a few updates.

**Comment by kshitij12345** [comment]
@mruberry 
Thanks for the great review with summary!

I believe have addressed the comments. Please review.

File: docs/source/name_inference.rst [hunk 5]
```diff
@@ -235,6 +235,8 @@ If you don't see an operation listed here, but it would help your use case, plea
    :func:`torch.var_mean`,:ref:`removes_dimensions-doc`
    :meth:`Tensor.zero_`,None
    :func:`torch.zeros`,:ref:`factory-doc`
+   :func:`torch.count_nonzero`, See documentation
```
Comment (mruberry) [inline docs/source/name_inference.rst:238]: Shouldn't this be :ref:`removes_dimensions-doc`?

File: test/test_torch.py [hunk 6]
```diff
@@ -17596,6 +17596,29 @@ def test_multinomial_empty(self, device):
             out = torch.multinomial(probs, num_samples=num_samples, replacement=replacement)
             self.assertEqual(out, expected)
 
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    @dtypes(*(torch.testing.get_all_int_dtypes() + torch.testing.get_all_fp_dtypes(include_bfloat16=False)))
+    def test_count_nonzero(self, device, dtype):
+        for ndims in (0, 2, 1, 3):
```
Comment (mruberry) [inline test/test_torch.py:17602]: nit: 0, 1, 2, 3?

File: test/test_torch.py [hunk 7]
```diff
@@ -17596,6 +17596,29 @@ def test_multinomial_empty(self, device):
...
+            shape = self._rand_shape(ndims, 5, 10)
+            for n in range(ndims + 1):
+                for c in combinations(list(range(ndims)), n):
+                    for count_dim in permutations(c):
+                        if shape == ():
+                            x = torch.tensor(())
```
Comment (mruberry) [inline test/test_torch.py:17608]: When constructing the tensor you still need to pass device=device and dtype=dtype

File: test/test_torch.py [hunk 8]
```diff
@@ -17596,6 +17596,29 @@ def test_multinomial_empty(self, device):
...
+                        else:
+                            x = torch.randn(*shape)
```
Comment (mruberry) [inline test/test_torch.py:17610]: You also need to construct this with the correct device and dtype. In addition, if the dtype is not a floating type you should use randint:

```
if not dtype.is_floating_point:
  t = torch.randint(1, 2, shape, device=device, dtype=dtype)
else:
  t = torch.randn(*shape, device=device).to(dtype)  
```

In the latter case I believe you'll need to .to the result of randn because I don't think we support randn for all float types on the CPU or CUDA.

Comment (mruberry) [reply test/test_torch.py:17610]: Btw, I took a note to simplify self.compare_with_numpy() so it's clearer how to use it. Sorry about that.

File: aten/src/ATen/native/native_functions.yaml [hunk 9]
```diff
@@ -805,6 +805,17 @@
...
 
+- func: count_nonzero.dim_IntList(Tensor self, int[] dim) -> Tensor
+  use_c10_dispatcher: full
+  variants: function, method
+
+- func: count_nonzero(Tensor self, int? dim=None) -> Tensor
+  use_c10_dispatcher: full
+  variants: function, method
+
+- func: count_nonzero.dim_DimnameList(Tensor self, Dimname[] dim) -> Tensor
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:816]: I would remove this variant, otherwise you'll have to add another test for it. It's not worth the hassle.

**Review by mruberry — CHANGES_REQUESTED** [review]
A few more test changes (sorry about that, I made a note to make self.compare_with_numpy easier to use!) and I suggest you remove the name variant to keep this PR more focused, otherwise you'll need to add tests for it, and that could be a separate PR if you wanted.

Also I made a note on the name_inference.rst doc.

File: test/test_torch.py (hunk 7 above)
Comment (kshitij12345) [reply test/test_torch.py:17608]: Now `x` is passed as `.tolist()` which will force it to take the else which considers `dtype` and `device`.

https://github.com/pytorch/pytorch/blob/18fe9d267c269851ba616f4e47120e5306c9827b/torch/testing/_internal/common_utils.py#L867-L873

File: aten/src/ATen/native/native_functions.yaml (hunk 9 above)
Comment (kshitij12345) [reply aten/src/ATen/native/native_functions.yaml]: Agreed. 
I was also having trouble with it. In Python, it wasn't correctly dispatching to the named variant when I tried to add test.

Have removed this as well as the mention in `namedinference.rst`

File: docs/source/name_inference.rst (hunk 5 above)
Comment (kshitij12345) [reply docs/source/name_inference.rst]: Have removed this, as we are not implementing the named variants here.

File: test/test_torch.py (hunk 7 above)
Comment (mruberry) [reply test/test_torch.py:17608]: Also there's one thing I failed to ask earlier: will NaN and inf work as expected? Is it worth adding a small separate test that just tests values (NaN, +inf, -inf) and makes sure we handle these extremals like NumPy does?

**Comment by mruberry** [comment]
https://github.com/pytorch/pytorch/pull/40064 clarifies the compare_with_numpy behavior, thank you for revealing an issue with it. You will have to rebase after it goes in, unfortunately, but the code change it requires is very small. We really want to avoid those tolist() calls.

File: test/test_torch.py (hunk 7 above)
Comment (kshitij12345) [reply test/test_torch.py:17608]: Right makes sense.

**Comment by kshitij12345** [comment]
@mruberry 

PTAL:)

**Comment by kshitij12345** [comment]
@mruberry 

Would be great if you can have another look at this. I am planning to use the `_generate_input` and `_test_reduction_function_with_numpy` for few tests for `nansum`.

Thank You.

File: test/test_torch.py [hunk 10]
```diff
@@ -17596,6 +17596,48 @@ def test_multinomial_empty(self, device):
...
+                            # Default `dims=None` case
+                            self.compare_with_numpy(torch_func, np_func, x, device=None, dtype=None)
+                        else:
+                            # With `dims: tuple of ints` case
+                            torch_func_partial = partial(torch_func, dim=count_dim)
+                            np_func_partial = partial(np_func, axis=count_dim)
+                            self.compare_with_numpy(torch_func_partial, np_func_partial, x, device=None, dtype=None)
+
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    @dtypes(*(torch.testing.get_all_int_dtypes() + torch.testing.get_all_fp_dtypes(include_bfloat16=False)))
```
Comment (mruberry) [inline test/test_torch.py:17636]: Should this include complex types?

**Review by mruberry — APPROVED** [review]
Excellent work. This looks really good, @kshitij12345!

Should the test also consider complex dtypes?

File: test/test_torch.py (hunk 10 above)
Comment (kshitij12345) [reply test/test_torch.py:17636]: Good point. I'll have to confirm that. If `masking` and `ne` support that.

Will confirm and get back.

Thanks.

Comment (kshitij12345) [reply test/test_torch.py]: Complex is supported. Have added support for complex test.

Thanks!

**Comment by kshitij12345** [comment]
@mruberry 

Have added complex dtypes to the test as well. 
I guess that should cover all the bases.

**Comment by mruberry** [comment]
> @mruberry
> 
> Have added complex dtypes to the test as well.
> I guess that should cover all the bases.

Thanks kshitij12345! We just need to wait for the tests again since there was a merge conflict with master.

**Comment by mruberry** [comment]
Another merge conflict, unfortunately. We'll have to wait for the tests again.

**Comment by kshitij12345** [comment]
@mruberry Gentle ping :)

**Comment by mruberry** [comment]
> @mruberry Gentle ping :)

Thanks for the ping. I'm fighting with our internal landing system, unfortunately.

**Comment by kshitij12345** [comment]
Oh. Sure. Thank You! :)
